src/platform/file_manager.c

```diff
@@ -0,0 +1,220 @@
+#include "file_manager.h"
+
+#include "core/dir.h"
+#include "core/file.h"
+#include "core/log.h"
+#include "core/string.h"
+#include "platform/vita/vita.h"
+
+#include <dirent.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#ifdef _WIN32
+#include <windows.h>
+
+#define fs_dir_type _WDIR
+#define fs_dir_entry struct _wdirent
+#define fs_dir_open _wopendir
+#define fs_dir_close _wclosedir
+#define fs_dir_read _wreaddir
+#define dir_entry_name(d) wchar_to_utf8(d->d_name)
+#define dir_entry_close_name(n) free((void*)n)
+
+static const char *wchar_to_utf8(const wchar_t *str)
+{
+    int size_needed = WideCharToMultiByte(CP_UTF8, 0, str, -1, NULL, 0, NULL, NULL);
+    char *result = (char *) malloc(sizeof(char) * size_needed);
+    WideCharToMultiByte(CP_UTF8, 0, str, -1, result, size_needed, NULL, NULL);
+    return result;
+}
+
+static wchar_t *utf8_to_wchar(const char *str)
+{
+    int size_needed = MultiByteToWideChar(CP_UTF8, 0, str, -1, NULL, 0);
+    wchar_t *result = (wchar_t *) malloc(sizeof(wchar_t) * size_needed);
+    MultiByteToWideChar(CP_UTF8, 0, str, -1, result, size_needed);
+    return result;
+}
+
+#else // not _WIN32
+#define fs_dir_type DIR
+#define fs_dir_entry struct dirent
+#define fs_dir_open opendir
+#define fs_dir_close closedir
+#define fs_dir_read readdir
+#define dir_entry_name(d) ((d)->d_name)
+#define dir_entry_close_name(n)
+#endif
+
+#ifndef S_ISLNK
+#define S_ISLNK(m) 0
+#endif
+
+#ifndef S_ISSOCK
+#define S_ISSOCK(m) 0
+#endif
+
+#ifdef __vita__
+#define CURRENT_DIR VITA_PATH_PREFIX
+#elif defined(_WIN32)
+#define CURRENT_DIR L"."
+#else
+#define CURRENT_DIR "."
+#endif
+
+#ifdef _MSC_VER
+#include <direct.h>
+#define chdir _chdir
+#elif !defined(__vita__)
+#include <unistd.h>
+#endif
+
+int platform_file_manager_get_directory_contents_by_extension(char ***file_list_ptr, int *file_count, const char *extension, int max_files)
```

Comment:
Nice idea.

To make things easier for you in the language switcher and avoid function repetition, I'm thinking the file and directory list filtering should happen in `platform_file_manager_list_files_in_directory`.

It would have the following parameters:

`void (*callback)(const char *), int type, const char *extension`

`type` would be a flag that could be set to `TYPE_DIR`, `TYPE_FILE` or `TYPE_ANY`. The listing would be filtered based on the type.

`extension` is an obvious one, but it would only work for files, not folders.

If needed, the base directory to list the files can also be added as a parameter.

Reply:
Good idea.

Nitpicks:
- change the parameter order to have the callback last: `int type, const char *extension, void (*callback)(const char *)`. It feels more logical to me: first specify which files you want, then specify what you want to do with them. Extension can be `NULL` for no filtering.
- since the function then also lists subdirs, rename it to just: `platform_file_manager_list_directory`

Reply:
I'm just here to say that these shouldn't be considered as `nitpicks` at all. These suggestions are obviously part of the clean coding principles.

Reply:
Haha I had already changed `callback` to the end before reading this 😉  It IS better that way!